external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala

```diff
@@ -171,9 +171,15 @@ private[sql] object AvroUtils extends Logging {
     protected val stopPosition: Long
 
     private[this] var completed = false
+    private[this] var interveningNext = true
+    private[this] var prevHasNextRow = false
     private[this] var currentRow: Option[InternalRow] = None
 
     def hasNextRow: Boolean = {
+      if (!interveningNext) {
+        // until a row is consumed, return previous result of hasNextRow
+        return prevHasNextRow
+      }
```

Comment:
Can't we just reset `currentRow` in  `nextRow` and check `currentRow.isDefined` here?

Reply:
I also feel @viirya 's suggestion would be simpler.

In addition, looks like the implementation didn't respect the Iterator's contracts - calling `hasNextRow` shouldn't be prerequisite to call `nextRow`.

Below code would fix the original issue (as this code passes the new test), as well as it would also work for the case only `nextRow` is called with handling NoSuchElementException.

```
def hasNextRow: Boolean = {
      if (currentRow.isDefined) {
        return true
      }

      while (!completed && currentRow.isEmpty) {
        val r = fileReader.hasNext && !fileReader.pastSync(stopPosition)
        if (!r) {
          fileReader.close()
          completed = true
          currentRow = None
        } else {
          val record = fileReader.next()
          currentRow = deserializer.deserialize(record).asInstanceOf[Option[InternalRow]]
        }
      }

      currentRow.isDefined
    }

    def nextRow: InternalRow = {
      if (currentRow.isEmpty) {
        if (!hasNextRow) {
          throw new NoSuchElementException("next on empty iterator")
        }
      }

      val row = currentRow.get
      currentRow = None
      row
    }
```

Reply:
I had the same thought initially but then I realised that @bersprockets wanted to do the least aggressive way. For example, previously we could get the same row from the next call of `nextRow`. After the approach above, we cannot although I think it's fine. I don't mind either way.

Reply:
Yeah, this approach is better, because it also handles the case where nextRow is called multiple times (which previously, would return the same row over and over).

Reply:
From the API semantics perspective, shouldn't `nextRow` return the next row? It looks okay if `hasNextRow` has been called multiple times before `nextRow` is called. But it sounds weird that `nextRow` will be called with the same row. As a fix this looks fine, but the API, if it is called like that way, sounds a weird design, in particular it is documented as iterator-like interface.

Reply:
>But it sounds weird that nextRow will be called with the same row.

Yes, I agree. As I mentioned above, leaving it that way would keep the status quo, but probably wouldn't be correct since the users of this RowReader are Iterator implementations.

Reply:
Follow up to my previous comment: As @HyukjinKwon pointed out, I was earlier trying to fix the stated bug without changing other behavior. But I should probably fix nextRow while I am at it, which makes RowReader follow a more recognized pattern and makes the actual Iterators that use it more correct.